pkg/controller/keyvault/nginx_secret_provider_class.go

```diff
@@ -0,0 +1,223 @@
+	var err error
+	result := ctrl.Result{}
+
+	// do metrics
+	defer func() {
+		//placing this call inside a closure allows for result and err to be bound after Reconcile executes
+		//this makes sure they have the proper value
+		//just calling defer metrics.HandleControllerReconcileMetrics(controllerName, result, err) would bind
+		//the values of result and err to their zero values, since they were just instantiated
+		metrics.HandleControllerReconcileMetrics(nginxSecretProviderControllerName, result, err)
+	}()
+
+	logger, err := logr.FromContext(ctx)
+	if err != nil {
+		return result, err
+	}
+	logger = nginxSecretProviderControllerName.AddToLogger(logger).WithValues("name", req.Name, "namespace", req.Namespace)
+
+	logger.Info("getting Nginx Ingress")
+	nic := &approutingv1alpha1.NginxIngressController{}
+	err = i.client.Get(ctx, req.NamespacedName, nic)
+	if err != nil {
+		return result, client.IgnoreNotFound(err)
+	}
+	logger = logger.WithValues("name", nic.Name, "namespace", "app-routing-system", "generation", nic.Generation)
+
+	spc := &secv1.SecretProviderClass{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "secrets-store.csi.x-k8s.io/v1",
+			Kind:       "SecretProviderClass",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      DefaultNginxCertName(nic),
+			Namespace: "app-routing-system",
+			Labels:    manifests.GetTopLevelLabels(),
+			OwnerReferences: []metav1.OwnerReference{{
+				APIVersion: nic.APIVersion,
+				Controller: util.BoolPtr(true),
+				Kind:       nic.Kind,
+				Name:       nic.Name,
+				UID:        nic.UID,
+			}},
+		},
+	}
+	logger = logger.WithValues("spc", spc.Name)
+	ok, err := i.buildSPC(nic, spc)
+	if err != nil {
+		logger.Info("failed to build secret provider class for ingress, user input invalid. sending warning event")
+		i.events.Eventf(nic, "Warning", "InvalidInput", "error while processing Keyvault reference: %s", err)
+		return result, nil
+	}
+	if ok {
+		logger.Info("reconciling secret provider class for ingress")
+		err = util.Upsert(ctx, i.client, spc)
+		if err != nil {
+			i.events.Eventf(nic, "Warning", "FailedUpdateOrCreateSPC", "error while creating or updating SecretProviderClass needed to pull Keyvault reference: %s", err)
+		}
+		return result, err
+	}
+
+	logger.Info("cleaning unused managed spc for ingress")
+	logger.Info("getting secret provider class for ingress")
+
+	toCleanSPC := &secv1.SecretProviderClass{}
+
+	err = i.client.Get(ctx, client.ObjectKeyFromObject(spc), toCleanSPC)
+	if err != nil {
+		return result, client.IgnoreNotFound(err)
+	}
+
+	if manifests.HasTopLevelLabels(toCleanSPC.Labels) {
+		logger.Info("removing secret provider class for ingress")
+		err = i.client.Delete(ctx, toCleanSPC)
+		return result, client.IgnoreNotFound(err)
+	}
+
+	return result, nil
+}
+
+func (i *NginxSecretProviderClassReconciler) buildSPC(nic *approutingv1alpha1.NginxIngressController, spc *secv1.SecretProviderClass) (bool, error) {
```

Comment:
this should likely be a more generic function. 

it's almost exactly the same as here

https://github.com/Azure/aks-app-routing-operator/blob/aa4f99d481269d43a42df4f9e8ee785b60ab394c/pkg/controller/keyvault/ingress_secret_provider_class.go#L140

we should be able to merge them together and use the same fn for both